src/sources/ping/eventfd.rs

```diff
@@ -0,0 +1,189 @@
+        // wake up the loop.
+        Err(nix::errno::Errno::EAGAIN) => Ok(()),
+
+        // Anything else is a real error.
+        Err(e) => Err(e.into()),
+    }
+}
+
+#[inline]
+fn drain_ping(fd: RawFd) -> std::io::Result<u64> {
+    // The eventfd counter is effectively a u64.
+    const NBYTES: usize = 8;
+    let mut buf = [0u8; NBYTES];
+
+    match read(fd, &mut buf) {
+        // Reading from an eventfd should only ever produce 8 bytes. No looping
+        // is required.
+        Ok(NBYTES) => Ok(u64::from_ne_bytes(buf)),
+
+        Ok(_) => unreachable!(),
+
+        // Any other error can be propagated.
+        Err(e) => Err(e.into()),
+    }
+}
+
+// The event source is simply a generic source with one of the eventfds.
+#[derive(Debug)]
+pub struct PingSource {
+    event: Generic<RawFd>,
+
+    // This is held only to ensure that there is an owner of the fd that lives
+    // as long as the Generic source, so that the fd is not closed unexpectedly
+    // when all the senders are dropped.
+    _fd: Arc<CloseOnDrop>,
+}
+
+impl EventSource for PingSource {
+    type Event = ();
+    type Metadata = ();
+    type Ret = ();
+    type Error = PingError;
+
+    fn process_events<C>(
+        &mut self,
+        readiness: Readiness,
+        token: Token,
+        mut callback: C,
+    ) -> Result<PostAction, Self::Error>
+    where
+        C: FnMut(Self::Event, &mut Self::Metadata) -> Self::Ret,
+    {
+        self.event
+            .process_events(readiness, token, |_, &mut fd| {
+                let counter = drain_ping(fd)?;
+
+                // If the LSB is set, it means we were closed. If anything else
+                // is also set, it means we were pinged. The two are not
+                // mutually exclusive.
+                let close = (counter & INCREMENT_CLOSE) != 0;
+                let ping = (counter & (u64::MAX - 1)) != 0;
+
+                if ping {
+                    callback((), &mut ());
+                }
+
+                if close {
+                    Ok(PostAction::Remove)
+                } else {
+                    Ok(PostAction::Continue)
+                }
+            })
+            .map_err(|e| PingError(e.into()))
+    }
+
+    fn register(&mut self, poll: &mut Poll, token_factory: &mut TokenFactory) -> crate::Result<()> {
+        self.event.register(poll, token_factory)
+    }
+
+    fn reregister(
+        &mut self,
+        poll: &mut Poll,
+        token_factory: &mut TokenFactory,
+    ) -> crate::Result<()> {
+        self.event.reregister(poll, token_factory)
+    }
+
+    fn unregister(&mut self, poll: &mut Poll) -> crate::Result<()> {
+        self.event.unregister(poll)
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Ping {
+    // This is an Arc because it's potentially shared with clones. The last one
+    // dropped needs to signal to the event source via the eventfd.
+    event: Arc<FlagOnDrop>,
```

Comment:
There is no need for nested arcs? The FlagOnDrop is already an arc? Otherwise whats the purpose of nesting arcs?

Reply:
If you clone the `FlagOnDrop` throughout all the senders, a different one will be dropped every time a sender is dropped. This will (a) wake the source up every time with `send_ping()` in its drop handler and (b) hit the problem you mentioned elsewhere, where the counter may be incremented by 2 (or an even number) instead of 1 if two are dropped in between loop dispatches.

There must *strictly* be one `FlagOnDrop` in existence for any given `PingSource` for the signalling to work.

To put it another way: `Arc`s are for shared ownership.
- The inner level of `Arc` is for shared ownership of the eventfd by both the `PingSource` and all `Ping`s, original or cloned.
- The outer level of `Arc` is for shared ownership of the *single* "thing that adds 1 to the eventfd counter" amongst all the `Ping` clones.

They do not conceptually share ownership of the same thing, so you need two levels.

Reply:
Here's a quick picture. Note how the top `Arc` (the one containing `FlagOnDrop`) is not shared with `PingSource`. This makes `FlagOnDrop` uniquely responsible for adding numbers to the eventfd counter.

![arcs](https://user-images.githubusercontent.com/311239/161535416-46228bbb-58b0-4ca7-967e-bda1d13ca999.png)